File: Play/NBKit/NBDial.m
//
//  NBDial.m
//  Play
//
//  Created by Nathan Borror on 4/19/13.
//  Copyright (c) 2013 Nathan Borror. All rights reserved.
//

#import "NBDial.h"
#import "NBDirectionGestureRecognizer.h"

@interface NBDial ()
{
  CGPoint panCoordBegan;
  UIImageView *max;
  UIView *min;

  CGFloat maxOriginX;
  CGFloat minOriginX;
}
@end

@implementation NBDial
@synthesize maxValue, minValue, value;

- (id)initWithFrame:(CGRect)frame
{
  self = [super initWithFrame:frame];
  if (self) {
    [self setClipsToBounds:YES];

    max = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, CGRectGetWidth(self.bounds), 25)];
    [max setImage:[[UIImage imageNamed:@"NBDial"] resizableImageWithCapInsets:UIEdgeInsetsMake(12, 150, 12, 150) resizingMode:UIImageResizingModeStretch]];
    [self addSubview:max];

    min = [[UIView alloc] initWithFrame:CGRectMake(0, 0, CGRectGetWidth(self.bounds), 25)];
    [min setClipsToBounds:YES];
    [self addSubview:min];

    UIView *minNotch = [[UIView alloc] initWithFrame:CGRectMake(CGRectGetWidth(min.bounds)-2, 0, 2, CGRectGetHeight(self.bounds))];
    [minNotch setBackgroundColor:[UIColor colorWithRed:0 green:0 blue:0 alpha:.1]];
    [min addSubview:minNotch];

    NBDirectionGestureRecognizer *pan = [[NBDirectionGestureRecognizer alloc] initWithTarget:self action:@selector(panDial:)];
    [pan setDirection:NBDirectionPanGestureRecognizerHorizontal];
    [self addGestureRecognizer:pan];

    maxOriginX = min.center.x-12;
    minOriginX = -(maxOriginX);
  }
  return self;
}

- (void)setValue:(CGFloat)aValue
{
  self->value = aValue;

  CGPoint newPoint = CGPointMake([self findPosition:aValue], min.center.y);

  if (newPoint.x < maxOriginX && newPoint.x > minOriginX) {
    min.center = newPoint;
  } else {
    if (newPoint.x > maxOriginX) {
      min.center = CGPointMake(maxOriginX, min.center.y);
    }
    if (newPoint.x < minOriginX) {
      min.center = CGPointMake(minOriginX, min.center.y);
    }
  }
}

#pragma mark - NBDirectionGestureRecognizer

- (void)panDial:(NBDirectionGestureRecognizer *)recognizer
{
  if (recognizer.state == UIGestureRecognizerStateBegan) {
    panCoordBegan = [recognizer locationInView:min];
  }

  if (recognizer.state == UIGestureRecognizerStateChanged) {
    CGPoint panCoordChange = [recognizer locationInView:min];
    CGFloat deltaX = panCoordChange.x - panCoordBegan.x;
    CGFloat newX = min.center.x + deltaX;

    // Figure out value
    [self setValue:[self findValue:newX]];
    [super sendActionsForControlEvents:UIControlEventValueChanged];
  }
}

- (CGFloat)findValue:(CGFloat)x
{
  return ((((maxValue - minValue) * (x - minOriginX)) / (maxOriginX - minOriginX)) + minValue);
}

- (CGFloat)findPosition:(CGFloat)x
{
  return ((((maxOriginX - minOriginX) * (x - minValue)) / (maxValue - minValue)) + minOriginX);
}

@end

File: Play/Custom/NBDial.swift
//
//  NBDial.swift
//  Play
//
//  Created by Nathan Borror on 6/14/14.
//  Copyright (c) 2014 Nathan Borror. All rights reserved.
//

import UIKit

class NBDial: UIControl {

    var panCoordBegan = CGPoint(x: 0.0, y: 0.0)
    var maxOrigin:CGFloat = 0.0
    var minOrigin:CGFloat = 0.0

    var maxValue:CGFloat = 0.0
    var minValue:CGFloat = 0.0
    var value: CGFloat = 0.0 {
        didSet {
            let newPoint = CGPoint(x: self.findPosition(value), y: thumb.center.y)
            if newPoint.x < maxOrigin && newPoint.x > minOrigin {
                thumb.center = newPoint;
            } else {
                if newPoint.x > maxOrigin {
                    thumb.center = CGPoint(x: maxOrigin, y: thumb.center.y);
                }
                if (newPoint.x < minOrigin) {
                    thumb.center = CGPoint(x: minOrigin, y: thumb.center.y);
                }
            }
            min.frame = CGRect(x: 0, y: 0, width: CGRectGetMaxX(thumb.frame), height: CGRectGetHeight(self.bounds))
        }
    }

    let max = UIView()
    let min = UIView()
    let thumb = UIView()

    init(frame: CGRect) {
        super.init(frame: frame)
        self.clipsToBounds = false

        max.frame = CGRect(x: 0.0, y: 0.0, width: CGRectGetWidth(self.frame), height: CGRectGetHeight(self.frame))
        max.backgroundColor = UIColor(patternImage: UIImage(named: "DialBackground"))
        max.autoresizingMask = .FlexibleWidth | .FlexibleHeight
        self.addSubview(max)

        min.frame = CGRect(x: 0.0, y: 0.0, width: 8.0, height: CGRectGetHeight(self.frame))
        min.backgroundColor = UIColor.tintColor()
        min.autoresizingMask = .FlexibleHeight
        self.addSubview(min)

        thumb.frame = CGRect(x: 0.0, y: 0.0, width: 4.0, height: CGRectGetHeight(self.frame))
        thumb.autoresizingMask = .FlexibleHeight
        self.addSubview(thumb)

        let panGesture = UIPanGestureRecognizer(target: self, action: "handlePanGesture:")
        self.addGestureRecognizer(panGesture)

        maxOrigin = CGRectGetWidth(self.bounds) - 15.0
        minOrigin = 15.0
    }

    override func layoutSubviews() {
        maxOrigin = CGRectGetWidth(self.bounds) - 15.0
    }

    func handlePanGesture(recognizer: UIPanGestureRecognizer) {
        if recognizer.state == UIGestureRecognizerState.Began {
            panCoordBegan = recognizer.locationInView(thumb)
        }

        if recognizer.state == UIGestureRecognizerState.Changed {
            let panCoordChange = recognizer.locationInView(thumb)
            let xDelta = panCoordChange.x - panCoordBegan.x
            let x = thumb.center.x + xDelta

            value = findValue(x)
            super.sendActionsForControlEvents(UIControlEvents.ValueChanged)
        }
    }

    func findValue(x: CGFloat) -> CGFloat {
        return ((((maxValue - minValue) * (x - minOrigin)) / (maxOrigin - minOrigin)) + minValue)
    }

    func findPosition(x: CGFloat) -> CGFloat {
        return ((((maxOrigin - minOrigin) * (x - minValue)) / (maxValue - minValue)) + minOrigin)
    }
}
